lib/puppet/etc.rb

```diff
@@ -53,70 +79,77 @@ def setpwent
     # returns an Etc::Passwd struct corresponding to the entry or raises
     # ArgumentError if none
     def getpwnam(username)
-      convert_field_values_to_utf8!(::Etc.getpwnam(username))
+      override_field_values_to_utf8(::Etc.getpwnam(username))
     end
 
     # Etc::getgrnam searches /etc/group file for an entry corresponding to groupname.
     # returns an Etc::Group struct corresponding to the entry or raises
     # ArgumentError if none
     def getgrnam(groupname)
-      convert_field_values_to_utf8!(::Etc.getgrnam(groupname))
+      override_field_values_to_utf8(::Etc.getgrnam(groupname))
     end
 
     # Etc::getgrid searches /etc/group file for an entry corresponding to id.
     # returns an Etc::Group struct corresponding to the entry or raises
     # ArgumentError if none
     def getgrgid(id)
-      convert_field_values_to_utf8!(::Etc.getgrgid(id))
+      override_field_values_to_utf8(::Etc.getgrgid(id))
     end
 
     # Etc::getpwuid searches /etc/passwd file for an entry corresponding to id.
     # returns an Etc::Passwd struct corresponding to the entry or raises
     # ArgumentError if none
     def getpwuid(id)
-      convert_field_values_to_utf8!(::Etc.getpwuid(id))
+      override_field_values_to_utf8(::Etc.getpwuid(id))
     end
 
-    private
-    # Utility method for converting the String values of a struct returned by
+     private
+
+    # @api private
+    # Defines Puppet::Etc::Passwd struct class. Contains all of the original
+    # member fields of Etc::Passwd, and additional "canonical_" versions of
+    # these fields as well. API compatible with Etc::Passwd. Because Struct.new
+    # defines a new Class object, we memozie to avoid superfluous extra Class
+    # instantiations.
+    def puppet_etc_passwd_class
+      @password_class ||= Struct.new(*Etc::Passwd.members, *Etc::Passwd.members.map { |member| "canonical_#{member}".to_sym })
```

Comment:
Nope, not particularly useful. To avoid that, we could hard code the list of available attributes here? I thought about it but decided to go with this because there's a couple things that make that more annoying
- the list varies based on the struct class (and knowing ruby could change without warning) so we maintain different lists based on if we're doing a user or group struct
- the list actually varies based on the flags passed to [ruby during compilation](https://ruby-doc.org/stdlib-2.3.0/libdoc/etc/rdoc/Etc.html#label-The+following+members+below+are+optional-2C+and+must+be+compiled+with+special+flags-3A) . This means we'd need to conditionally check for the presence of the members that we expect to exist and have string values and only duplicate those. I feel like there's a tradeoff there in maintainability vs the completeness of just duplicating all the members
- A sidenote that doesn't necessarily impact this discussion but is worth noting is that the Etc struct members aren't actually typed - any field could just as easily return a string as an integer. So if for some reason the underlying system changes the return value for any of these we're going to be in trouble. We could guard by only trying to encode things that are strings rather than expecting specific fields to be, but again that felt like going to far. Happy to entertain alternatives there though.